src/main/java/org/broadinstitute/hellbender/utils/read/ArtificialBAMBuilder.java

```diff
@@ -0,0 +1,222 @@
+        Utils.validateArg(nLoci > 0, "nLoci should be positive but was " + nLoci);
+        this.nReadsPerLocus = nReadsPerLocus;
+        this.nLoci = nLoci;
+        this.reference = null;
+        this.parser = new GenomeLocParser(dict);
+        createAndSetHeader(1);
+    }
+
+    public IndexedFastaSequenceFile getReference() {
+        return reference;
+    }
+
+    public GenomeLocParser getGenomeLocParser() {
+        return parser;
+    }
+
+    public ArtificialBAMBuilder createAndSetHeader(final int nSamples) {
+        Utils.validateArg(nSamples > 0, "nSamples should be positive but was " + nSamples);
+
+        createdReads = null;
+        this.header = new SAMFileHeader();
+        header.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+        header.setSequenceDictionary(parser.getSequenceDictionary());
+        samples.clear();
+
+        for ( int i = 0; i < nSamples; i++ ) {
+            final SAMReadGroupRecord rg = new SAMReadGroupRecord("rg" + i);
+            final String sample = "sample" + i;
+            samples.add(sample);
+            rg.setSample(sample);
+            rg.setPlatform(NGSPlatform.ILLUMINA.getDefaultPlatform());
+            header.addReadGroup(rg);
+        }
+
+        return this;
+    }
+
+    public List<String> getSamples() {
+        return samples;
+    }
+
+    /**
+     * Create a read stream based on the parameters.  The cigar string for each
+     * read will be *M, where * is the length of the read.
+     *
+     * Useful for testing things like LocusIteratorBystate
+     *
+     * @return a ordered list of reads
+     */
+    public List<GATKRead> makeReads() {
+        if ( createdReads == null ) {
+            final String baseName = "read";
+            final LinkedList<SAMReadGroupRecord> readGroups = new LinkedList<>();
+            for ( final SAMReadGroupRecord rg : header.getReadGroups()) {
+                readGroups.add(rg);
+            }
+
+            final List<GATKRead> reads = new ArrayList<>(nReadsPerLocus*nLoci);
+            for ( int locusI = 0; locusI < nLoci; locusI++) {
+                final int locus = locusI * (skipNLoci + 1);
+                for ( int readI = 0; readI < nReadsPerLocus; readI++ ) {
+                    for ( final SAMReadGroupRecord rg : readGroups ) {
+                        final String readName = String.format("%s.%d.%d.%s", baseName, locus, readI, rg.getId());
+                        final GATKRead read = ArtificialReadUtils.createArtificialRead(header, readName, 0, alignmentStart + locus, readLength);
+                        read.setReadGroup(rg.getId());
+                        reads.add(read);
+                    }
+                }
+            }
+
+            if ( ! additionalReads.isEmpty() ) {
+                reads.addAll(additionalReads);
+                Collections.sort(reads, new ReadCoordinateComparator(header));
```

Comment:
This comparator will not produce an order 100% consistent with the ordering expected in a bam file (in particular, it will put unmapped reads that have the position of their mates in with the rest of the unmapped reads). Best way to fix this is to set `preSorted` to false when creating your bam writer below and let the writer do the sorting.